lib/irb.rb

```diff
@@ -968,7 +972,8 @@ def run(conf = IRB.conf)
       conf[:IRB_RC].call(context) if conf[:IRB_RC]
       conf[:MAIN_CONTEXT] = context
 
-      save_history = !in_nested_session && conf[:SAVE_HISTORY] && context.io.support_history_saving?
+      supports_history_saving = conf[:SAVE_HISTORY] && context.io.support_history_saving?
+      save_history = !in_nested_session && supports_history_saving
```

Comment:
Why do we need to separate this?

Reply:
`save_history` is false when `in_nested_session` is true. When calling `exit!`, we still want to save the history even if the session is nested.

```console
binding.irb
binding.irb
exit! -> should save history
```

```console
binding.irb
binding.irb
exit -> should not save history
a = 3
b = 44
exit -> should save history
```

Reply:
I see, that's a good catch 👍 
Do you mind adding a test case for this too?

Reply:
It is covered by https://github.com/ruby/irb/pull/851/files#diff-c37538faf5a6aa1413dc2844e6e78c58656b6474985c895e743ed422be4b3875R278-R280. I can change the message to something more explicit if needed.